rust/datafusion/src/datasource/memory.rs

```diff
@@ -0,0 +1,163 @@
+//! queried by DataFusion. This allows data to be pre-loaded into memory and then repeatedly
+//! queried without incurring additional file I/O overhead.
+
+use std::cell::RefCell;
+use std::rc::Rc;
+use std::sync::Arc;
+
+use arrow::datatypes::Schema;
+use arrow::record_batch::RecordBatch;
+
+use crate::datasource::{RecordBatchIterator, Table};
+use crate::execution::error::Result;
+
+pub struct MemTable {
+    schema: Arc<Schema>,
+    batches: Vec<RecordBatch>,
+}
+
+impl MemTable {
+    pub fn new(schema: Arc<Schema>, batches: Vec<RecordBatch>) -> Self {
+        Self { schema, batches }
+    }
+
+    /// Create a mem table by reading from another data source
+    pub fn load(t: &Table) -> Self {
+        let schema = t.schema();
+        let it = t.scan(&None, 1024 * 1024);
+        let mut it_mut = it.borrow_mut();
+
+        let mut data: Vec<RecordBatch> = vec![];
+
+        while let Ok(Some(batch)) = it_mut.next() {
+            data.push(batch);
+        }
+        MemTable::new(schema.clone(), data)
+    }
+}
+
+impl Table for MemTable {
+    fn schema(&self) -> &Arc<Schema> {
+        &self.schema
+    }
+
+    fn scan(
+        &self,
+        projection: &Option<Vec<usize>>,
+        _batch_size: usize,
+    ) -> Rc<RefCell<RecordBatchIterator>> {
+        let columns: Vec<usize> = match projection {
+            Some(p) => p.clone(),
+            None => {
+                let l = self.schema.fields().len();
+                let mut v = Vec::with_capacity(l);
+                for i in 0..l {
+                    v.push(i);
+                }
+                v
+            }
+        };
+
+        let projected_schema = Arc::new(Schema::new(
+            columns
+                .iter()
+                .map(|i| self.schema.field(*i).clone())
+                .collect(),
+        ));
+
+        Rc::new(RefCell::new(MemBatchIterator {
+            schema: projected_schema.clone(),
+            index: 0,
+            batches: self
+                .batches
+                .iter()
+                .map(|batch| {
+                    RecordBatch::new(
+                        projected_schema.clone(),
+                        columns.iter().map(|i| batch.column(*i).clone()).collect(),
+                    )
+                })
+                .collect(),
+        }))
+    }
+}
+
+pub struct MemBatchIterator {
+    schema: Arc<Schema>,
+    index: usize,
+    batches: Vec<RecordBatch>,
+}
+
+impl RecordBatchIterator for MemBatchIterator {
+    fn schema(&self) -> &Arc<Schema> {
+        &self.schema
+    }
+
+    fn next(&mut self) -> Result<Option<RecordBatch>> {
+        if self.index < self.batches.len() {
+            self.index += 1;
+            Ok(Some(self.batches[self.index - 1].clone()))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use arrow::array::Int32Array;
+    use arrow::datatypes::{DataType, Field, Schema};
+
+    #[test]
+    fn test1() {
```

Comment:
I think this test name could be more descriptive than `test1`, maybe even split this into 2 tests, no projection and with projection?

Reply:
Done